Kubernetes v1.32 sneak peek blog

mbianchidev opened the pull request:
Add 2024-11-04-kubernetes-1.32-sneak-peek.md

Preview blog link: https://deploy-preview-48585--kubernetes-io-main-staging.netlify.app/blog/2024/11/08/kubernetes-1-32-upcoming-changes/

mbianchidev commented:
Preview of the blog: https://deploy-preview-48585--kubernetes-io-main-staging.netlify.app/blog/2024/11/04/kubernetes-1-32-upcoming-changes/

dipesh-rawat commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+This blog outlines some of the planned changes for the Kubernetes v1.32 release, that the release team feels you should be aware of, for the continued mantainance of your Kubernetes enviroment and keeping up to date with the latest changes. Information listed below is based on the current status of the v1.32 release and may change before the actual release date. 
+
+### The Kubernetes API Removal and Deprecation process
+The Kubernetes project has a well-documented [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) for features. This policy states that stable APIs may only be deprecated when a newer, stable version of that same API is available and that APIs have a minimum lifetime for each stability level. A deprecated API has been marked for removal in a future Kubernetes r ...
```
We use site-relative link where possible.

```suggestion
The Kubernetes project has a well-documented [deprecation policy](/docs/reference/using-api/deprecation-policy/) for features. This policy states that stable APIs may only be deprecated when a newer, stable version of that same API is available and that APIs have a minimum lifetime for each stability level. A deprecated API has been marked for removal in a future Kubernetes release, it will continue to function until removal (at least one year from the deprecation), but usage will result in a warning being displayed. Removed APIs are no longer available in the current version, at which point you must migrate to using the replacement.
```

dipesh-rawat commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+* Alpha or experimental API versions may be removed in any release without prior deprecation notice, this process can become a withdrawal in cases where a different implementation for the same feature is already in place.
+
+Whether an API is removed as a result of a feature graduating from beta to stable or because that API simply did not succeed, all removals comply with this deprecation policy. Whenever an API is removed, migration options are communicated in the [documentation](https://kubernetes.io/docs/reference/using-api/deprecation-guide/).
```
```suggestion
Whether an API is removed as a result of a feature graduating from beta to stable or because that API simply did not succeed, all removals comply with this deprecation policy. Whenever an API is removed, migration options are communicated in the [documentation](/docs/reference/using-api/deprecation-guide/).
```

atiratree commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Control over terminating Pods in Deployments - KEP [#3973](https://github.com/kubernetes/enhancements/issues/3973)
+
+This enhancement adds a new field `.spec.PodReplacementPolicy` to Deployments, to specify whether new Pods can be scheduled only after old Pods have finished terminating, or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can potentially help avoid problems with excessive resource usage or slow scheduling during roll ...
```
```suggestion
This enhancement adds a new field `.spec.podReplacementPolicy` to Deployments, to specify whether new Pods can be created only after old Pods have finished terminating, or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can help to better control the speed of rollouts/scaling, and potentially avoid problems with excessive resource usage.
```

I think it is better not to mention scheduling, as this is techincally done by the scheduler and not affected by the deployment controller.

mbianchidev replied:
Thank you @atiratree !

SergeyKanzhelev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+Currently, Kubernetes provides graceful node shutdown functionality for Linux nodes but lacked equivalent support for Windows. This enhancement aims to enable the kubelet on Windows nodes to handle system shutdown events properly. By doing so, it ensures that Pods running on Windows nodes are gracefully terminated, allowing workloads to be rescheduled without disruption. This improvement enhances ...
+
+KEP [#4885](https://github.com/kubernetes/enhancements/issues/4885) adds CPU and memory affinity support to Windows nodes in Kubernetes, thus improving performance for workloads that benefit from cache vicinity and NUMA optimizations, this effort also aligns Windows capabilities with Linux nodes.
+
```
A few more exciting things from SIG Node around quility of life and small improvements:

1. https://github.com/kubernetes/kubernetes/pull/127566 - support for systemd watchdog is implemented for the kubelet. 
2. Better message in image pull backoff message: https://github.com/kubernetes/kubernetes/pull/127918

Also we are preparing to the major GA in 1.33 - tracking last remaining items and feedback for sidecar containers KEP: https://github.com/kubernetes/enhancements/issues/753#issuecomment-2350136594

mbianchidev replied:
Hey @SergeyKanzhelev 👋 

Thanks for pointing these out, amazing enhancements _huge fan of the better image pull backoff error btw_
Just so you know this blog being the sneak peek means that the KEP list chosen to be highlighted is non-exhaustive by definition, it's an arbitrary subset by the Release Team (Enhancements, Comms).

You might want to add these to the [Release Highlights discussion](https://github.com/kubernetes/sig-release/discussions/2639) for SIG Node instead, those and any KEPs highlighted by the SIG leadership will be included in the final Release Blog!

SergeyKanzhelev replied:
sounds good. https://github.com/kubernetes/sig-release/discussions/2639#discussioncomment-11093925

mbianchidev replied:
Amazing, thank you!

divya-mohan0209 commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+This blog outlines some of the planned changes for the Kubernetes v1.32 release, that the release team feels you should be aware of, for the continued mantainance of your Kubernetes enviroment and keeping up to date with the latest changes. Information listed below is based on the current status of the v1.32 release and may change before the actual release date. 
+
+### The Kubernetes API Removal and Deprecation process
```
Please make this sentence case.

mbianchidev replied:
Solved!

divya-mohan0209 commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+In this release, like the previous one, the Kubernetes project continues to introduce enhancements to the Dynamic Resource Allocation (DRA), which is a key component of the Kubernetes resource management system. These enhancements aim to improve the flexibility and efficiency of resource allocation for workloads that require specialized hardware, such as GPUs, FPGAs and network adapters. The enha ...
+
+- Partionable devices support: KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815)
```
Spelling nit
```suggestion
- Partitionable devices support: KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815)
```

mbianchidev replied:
Thanks, fixed it!

divya-mohan0209 commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+KEP [#4885](https://github.com/kubernetes/enhancements/issues/4885) adds CPU and memory affinity support to Windows nodes in Kubernetes, thus improving performance for workloads that benefit from cache vicinity and NUMA optimizations, this effort also aligns Windows capabilities with Linux nodes.
+
+### Almost every ASCII character in env variables - KEP [#4369](https://github.com/kubernetes/enhancements/issues/4369)
```
I'd rephrase it to the below since the existing sentence sounds incomplete. WDYT?

```suggestion
### Allow special characters in environment variables - KEP [#4369](https://github.com/kubernetes/enhancements/issues/4369)
```

divya-mohan0209 commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Field status.hostIPs added for Pod - KEP [#2681](https://github.com/kubernetes/enhancements/issues/2681)
+
+This enhancement introduces a new field, status.hostIPs, to the Kubernetes Pod API, enabling support for multiple IP addresses (IPv4 and IPv6) assigned to a node. Previously, the Pod status only included a singular HostIP field, which limited address configurations, particularly in dual-stack networks. This enhancement allows the API to store multiple IP addresses in an array, accessible via the  ...
```
```suggestion
This enhancement introduces a new field, `status.hostIPs`, to the Kubernetes Pod API, enabling support for multiple IP addresses (IPv4 and IPv6) assigned to a node. Previously, the Pod status only included a singular HostIP field, which limited address configurations, particularly in dual-stack networks. This enhancement allows the API to store multiple IP addresses in an array, accessible via the Downward API, which applications can leverage for environments with complex IP requirements, such as IPv6 migrations. 
```

mbianchidev replied:
Another fix in, thank you!

divya-mohan0209 reviewed (commented):
Overarching feedback:

- Section title casing to be amended from KEP [ABCD](#) to [KEP ABCD](#)
- It'd be beneficial to add the one-line enhancement description as the title instead of reinventing the wheel, IMO.
- Casing for section titles should adhere to style guide.

mbianchidev commented:
> Overarching feedback:
> 
> * Section title casing to be amended from KEP [ABCD](#) to [KEP ABCD](#)
> * It'd be beneficial to add the one-line enhancement description as the title instead of reinventing the wheel, IMO.
> * Casing for section titles should adhere to style guide.

Hey @divya-mohan0209 thanks a lot for taking a look at this so quickly!
I amended the KEP titles and same goes for the section titles, can you take a look to see if everything looks fine now?

Regarding the one-line-enhancement description as title we decided not to do that for a couple of reasons:
- Sometimes the one line description is not that descriptive for people who might not have context; see the allowing all characters in env variables one;
- Adding some flavor (and pop culture) to this blog to make it a bit more fun to read;
- Creating extra sections having an umbrella title and subtitles like the Windows and the DRA ones.

Of course we are happy to change the titles in case we need to.
Let me know!

danwinship commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Make Kubernetes aware of the LoadBalancer behavior - [KEP #1860](https://github.com/kubernetes/enhancements/issues/1860)
+
+KEP [#1860](https://github.com/kubernetes/enhancements/issues/1860) graduates to GA, introducing the `ipMode` field for a Service of `type: LoadBalancer`, which can be set to either `"VIP"` or `"Proxy"`. The existing behavior of kube-proxy is preserved when using `"VIP"`, where kube-proxy handles the load balancing. The use of `"Proxy"` results in traffic sent directly to the load balancer, provi ...
```
So the KEP is maybe not super clear about this, but this isn't an end-user feature; the feature is there so that _people implementing cloud providers_ can better describe how their load balancers work, so that kube-proxy can integrate with them better, and then the end-user experience is just "some kinds of load balancers work better on some clouds than they used to" but the end user doesn't actually have to do anything to get that improvement. (So from the end user perspective, this isn't a new feature, it's just a bugfix.)

(And I'm not sure what cloud providers needed this, and which ones have adopted it...)

mbianchidev replied:
Clarified adding some context for the end user, thanks for your contribution!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+**Add resource health status to the Pod status**
+
+When a Pod is using a device that has failed or is temporarily unhealthy, it is difficult to know. KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680) makes troubleshooting of Pod crashes easier by exposing device health via `Pod Status`.
```
```suggestion
It isn't easy to know when a Pod uses a device that has failed or is temporarily unhealthy. KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680) makes troubleshooting Pod crashes easier by exposing device health via `Pod Status`.
```

mbianchidev replied:
Implemented, thank you!

ArvindParekh commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+- Partitionable devices support: [KEP #4815](https://github.com/kubernetes/enhancements/issues/4815)
+- Resource Claim Status and standardized network interface data: [KEP #4817](https://github.com/kubernetes/enhancements/issues/4817)
+- Structured parameters: KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381)
```
```suggestion
- Structured parameters: [KEP #4381](https://github.com/kubernetes/enhancements/issues/4381)
```
To maintain consistency in KEP number formatting with other points

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Windows strikes back!
+
+KEP [#4802](https://github.com/kubernetes/enhancements/issues/4802) adds support for graceful shutdown of Windows nodes in Kubernetes clusters.
```
```suggestion
KEP [#4802](https://github.com/kubernetes/enhancements/issues/4802) adds support for the graceful shutdown of Windows nodes in Kubernetes clusters.
```

mbianchidev replied:
Implemented a slightly different change, thanks!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+### Windows strikes back!
+
+KEP [#4802](https://github.com/kubernetes/enhancements/issues/4802) adds support for graceful shutdown of Windows nodes in Kubernetes clusters.
+Currently, Kubernetes provides graceful node shutdown functionality for Linux nodes but lacked equivalent support for Windows. This enhancement aims to enable the kubelet on Windows nodes to handle system shutdown events properly. By doing so, it ensures that Pods running on Windows nodes are gracefully terminated, allowing workloads to be rescheduled without disruption. This improvement enhances ...
```
```suggestion
Currently, Kubernetes provides graceful node shutdown functionality for Linux nodes but lacks equivalent support for Windows. This enhancement aims to enable the kubelet on Windows nodes to handle system shutdown events properly. Doing so ensures that Pods running on Windows nodes are gracefully terminated, allowing workloads to be rescheduled without disruption. This improvement enhances the reliability and stability of clusters that include Windows nodes, especially during planned maintenance or system updates.
```

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+KEP [#4802](https://github.com/kubernetes/enhancements/issues/4802) adds support for graceful shutdown of Windows nodes in Kubernetes clusters.
+Currently, Kubernetes provides graceful node shutdown functionality for Linux nodes but lacked equivalent support for Windows. This enhancement aims to enable the kubelet on Windows nodes to handle system shutdown events properly. By doing so, it ensures that Pods running on Windows nodes are gracefully terminated, allowing workloads to be rescheduled without disruption. This improvement enhances ...
+
+KEP [#4885](https://github.com/kubernetes/enhancements/issues/4885) adds CPU and memory affinity support to Windows nodes in Kubernetes, thus improving performance for workloads that benefit from cache vicinity and NUMA optimizations, this effort also aligns Windows capabilities with Linux nodes.
```
```suggestion
KEP [#4885](https://github.com/kubernetes/enhancements/issues/4885) adds CPU and memory affinity support to Windows nodes in Kubernetes, thus improving performance for workloads that benefit from cache vicinity and NUMA optimizations; this effort also aligns Windows capabilities with Linux nodes.
```

mbianchidev replied:
Implemented, thanks!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+As we get closer to the release date for Kubernetes v1.32, the project develops and matures, features may be deprecated, removed, or replaced with better ones for the project's overall health. 
+
+This blog outlines some of the planned changes for the Kubernetes v1.32 release, that the release team feels you should be aware of, for the continued mantainance of your Kubernetes enviroment and keeping up to date with the latest changes. Information listed below is based on the current status of the v1.32 release and may change before the actual release date. 
```
Solved, thanks!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Allow special characters in environment variables - [KEP #4369](https://github.com/kubernetes/enhancements/issues/4369)
+
+With the graduation of KEP-4369 to beta, Kubernetes now allows almost all printable ASCII characters (excluding "=") to be used as environment variable names. This change addresses the limitations previously imposed on variable naming, facilitating broader adoption of Kubernetes by accommodating various application needs. The relaxed validation will be enabled by default via `RelaxedEnvironmentVa ...
```
```suggestion
With the graduation of KEP-4369 to beta, Kubernetes now allows almost all printable ASCII characters (excluding "=") to be used as environment variable names. This change addresses the limitations previously imposed on variable naming, facilitating the broader adoption of Kubernetes by accommodating various application needs. The relaxed validation will be enabled by default via the `RelaxedEnvironmentVariableValidation` feature gate, ensuring that users can easily utilize environment variables without strict constraints, enhancing flexibility for developers working with applications like .NET Core that require special characters in their configurations.
```

mbianchidev replied:
Implemented, thank you!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Make Kubernetes aware of the LoadBalancer behavior - [KEP #1860](https://github.com/kubernetes/enhancements/issues/1860)
+
+KEP [#1860](https://github.com/kubernetes/enhancements/issues/1860) graduates to GA, introducing the `ipMode` field for a Service of `type: LoadBalancer`, which can be set to either `"VIP"` or `"Proxy"`. The existing behavior of kube-proxy is preserved when using `"VIP"`, where kube-proxy handles the load balancing. The use of `"Proxy"` results in traffic sent directly to the load balancer, provi ...
```
```suggestion
KEP [#1860](https://github.com/kubernetes/enhancements/issues/1860) graduates to GA, introducing the `ipMode` field for a Service of `type: LoadBalancer`, which can be set to either `"VIP"` or `"Proxy"`. The existing behavior of kube-proxy is preserved when using `"VIP"`, where kube-proxy handles the load balancing. Using `"Proxy"` results in traffic sent directly to the load balancer, providing cloud providers greater control over relying on kube-proxy.
```

mbianchidev replied:
Implemented, thank you!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+KEP [#1860](https://github.com/kubernetes/enhancements/issues/1860) graduates to GA, introducing the `ipMode` field for a Service of `type: LoadBalancer`, which can be set to either `"VIP"` or `"Proxy"`. The existing behavior of kube-proxy is preserved when using `"VIP"`, where kube-proxy handles the load balancing. The use of `"Proxy"` results in traffic sent directly to the load balancer, provi ...
+
+### Support PSI based on cgroupv2 - [KEP #4205](https://github.com/kubernetes/enhancements/issues/4205)
+This KEP adds support in the Kubelet to read Pressure Stall Information (PSI) metrics for CPU, Memory and I/O resources exposed from cAdvisor and runc. Pressure metrics are like a barometer that can warn about impending resource shortages on a node. This will enable the Kubelet to report node conditions, which can be utilized to prevent scheduling of Pods on nodes experiencing significant resourc ...
```
```suggestion
This KEP adds support in the Kubelet to read Pressure Stall Information (PSI) metrics for CPU, Memory, and I/O resources exposed from cAdvisor and runc. Pressure metrics are like a barometer that can warn about impending resource shortages on a node. This will enable the Kubelet to report node conditions, which can be utilized to prevent the scheduling of Pods on nodes experiencing significant resource constraints.
```

mbianchidev replied:
Implemented, thank you!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+This KEP adds support in the Kubelet to read Pressure Stall Information (PSI) metrics for CPU, Memory and I/O resources exposed from cAdvisor and runc. Pressure metrics are like a barometer that can warn about impending resource shortages on a node. This will enable the Kubelet to report node conditions, which can be utilized to prevent scheduling of Pods on nodes experiencing significant resourc ...
+
+### Retry generate name for resources - [KEP #4420](https://github.com/kubernetes/enhancements/issues/4420)
+This enhancement improves the handling of name conflicts for Kubernetes resources created with the `generateName` field. Previously, if a name conflict occurred, the API server returned a 409 error, and clients had to manually retry the request. With this update, the API server now automatically retries generating a new name up to seven times in case of a conflict. This significantly reduces the  ...
```
```suggestion
This enhancement improves handling name conflicts for Kubernetes resources created with the `generateName` field. Previously, if a name conflict occurred, the API server returned a 409 error, and clients had to manually retry the request. With this update, the API server automatically retries, generating a new name up to seven times in case of a conflict. This significantly reduces the chances of collision, ensuring smooth generation of up to 1 million names with less than a 0.1% probability of a conflict, providing more resilience for large-scale workloads.
```

mbianchidev replied:
Solved with a slightly different change, thank you!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Mutating admission policies - [KEP #3962](https://github.com/kubernetes/enhancements/issues/3962)
+
+This enhancement introduces "Mutating Admission Policies" using CEL (Common Expression Language), allowing Kubernetes users to define in-line mutating policies without the need for complex admission webhooks. 
```
```suggestion
This enhancement introduces "Mutating Admission Policies" using CEL (Common Expression Language), allowing Kubernetes users to define in-line mutating policies without complex admission webhooks. 
```

mbianchidev replied:
Implemented, thanks!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+### Mutating admission policies - [KEP #3962](https://github.com/kubernetes/enhancements/issues/3962)
+
+This enhancement introduces "Mutating Admission Policies" using CEL (Common Expression Language), allowing Kubernetes users to define in-line mutating policies without the need for complex admission webhooks. 
+This enhancement simplifies common mutating tasks—such as adding labels, setting fields, or injecting sidecars—by enabling policies to be directly defined within the Kubernetes API server. With CEL's object instantiation and Server-Side Apply merge algorithms, these policies can minimize reinvocation, reduce latency, and eliminate the overhead associated with webhook management. This new approach ...
```
```suggestion
This enhancement simplifies common mutating tasks—such as adding labels, setting fields, or injecting sidecars—by enabling policies to be directly defined within the Kubernetes API server. With CEL's object instantiation and Server-Side Apply merge algorithms, these policies can minimize reinvocation, reduce latency, and eliminate the overhead associated with webhook management. This new approach enhances Kubernetes extensibility by reducing operational complexity. Also, it provides a scalable, built-in solution for secure and efficient policy management across diverse use cases like GitOps and CI/CD, making Kubernetes admission control more accessible and powerful for developers and administrators alike.
```

mbianchidev replied:
Solved with a slightly different change!

nate-double-u commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Control over terminating Pods in deployments - [KEP #3973](https://github.com/kubernetes/enhancements/issues/3973)
+
+This enhancement adds a new field `.spec.podReplacementPolicy` to Deployments, to specify whether new Pods can be created only after old Pods have finished terminating, or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can help to better control the speed of rollouts/scaling, and potentially avoid problems with exce ...
```
```suggestion
This enhancement adds a new field `.spec.podReplacementPolicy` to Deployments to specify whether new Pods can be created only after old Pods have finished terminating or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can better control the speed of rollouts/scaling and potentially avoid problems with excessive resource usage.
```

mbianchidev replied:
Implemented, thanks!

nate-double-u reviewed (commented):
A copy-edit and spelling pass. Feel free to take or leave any given suggestion. Also, I had to refresh mid-review; I've checked for consistency, but my apologies if anything seems misaligned.

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816) adds support for a prioritized list of selection criteria, attached to a device request in a `ResourceClaim`.
+The DRA's Structured Parameters feature has added the ability to make requests for very specific types of devices using a `ResourceClaim`.
+
+However, the current API did not allow the user to indicate any priority when multiple types or configurations of devices might have met the needs of the workload. This feature allows the user to specify alternative requests that satisfy the workloads need, giving more flexibility to the scheduler in the workloads' scheduling phase.
```
Implemented, thanks!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+The original dynamic resource allocation (DRA) uses claim and class parameters that are opaque to Kubernetes. KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) introduces structured parameters so that kube-scheduler and Cluster Autoscaler can handle a claim allocation themselves without relying on a third-party driver.
+As users are increasingly adopting Kubernetes as their management solution for batch-processing workloads and edge computing, the need of exposing specialized hardware to Pods is increasing.
+Such workloads no longer need just RAM and CPU - so this KEP introduced a new API for describing which of these specialized resources a Pod needs, for example devices like a GPU or other kinds of accelerators.
```
Implemented, thank you!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+**DRA Structured Parameters** 
+
+The original dynamic resource allocation (DRA) uses claim and class parameters that are opaque to Kubernetes. KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) introduces structured parameters so that kube-scheduler and Cluster Autoscaler can handle a claim allocation themselves without relying on a third-party driver.
+As users are increasingly adopting Kubernetes as their management solution for batch-processing workloads and edge computing, the need of exposing specialized hardware to Pods is increasing.
```
Implemented, thanks!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+**DRA Structured Parameters** 
+
+The original dynamic resource allocation (DRA) uses claim and class parameters that are opaque to Kubernetes. KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) introduces structured parameters so that kube-scheduler and Cluster Autoscaler can handle a claim allocation themselves without relying on a third-party driver.
```
Implemented, thank you!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815) introduces improvements to Kubernetes' Dynamic Resource Allocation (DRA) framework, enabling more efficient support for high-performance partitionable devices like GPUs, FPGAs, and network adapters. Currently, these devices are treated as whole units, potentially leading to resource inefficiencies when workloads require only a fr ...
+By extending the DRA framework, this enhancement allows workloads to request and consume portions of these resources, increasing utilization and scheduling flexibility. The proposed enhancement also provides primitives for representing both full devices and their partitions in a more compact manner, supporting the recent transition from what was called "classic DRA" to the "structured parameters" ...
+The newly proposed extensions, fully transparent to the end-user, enable vendors to provision “overlapping” partitions, ensuring that the scheduler does not allocate conflicting partitions. This approach allows vendors to dynamically create new partitions after allocation, eliminating the need for pre-partitioned configurations, while still supporting device partition selection through existing m ...
```
Implemented!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+**Partitionable devices support**
+
+KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815) introduces improvements to Kubernetes' Dynamic Resource Allocation (DRA) framework, enabling more efficient support for high-performance partitionable devices like GPUs, FPGAs, and network adapters. Currently, these devices are treated as whole units, potentially leading to resource inefficiencies when workloads require only a fr ...
+By extending the DRA framework, this enhancement allows workloads to request and consume portions of these resources, increasing utilization and scheduling flexibility. The proposed enhancement also provides primitives for representing both full devices and their partitions in a more compact manner, supporting the recent transition from what was called "classic DRA" to the "structured parameters" ...
```
Implemented!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+### Even more DRA enhancements!
+
+In this release, like the previous one, the Kubernetes project continues to introduce enhancements to the Dynamic Resource Allocation (DRA), which is a key component of the Kubernetes resource management system. These enhancements aim to improve the flexibility and efficiency of resource allocation for workloads that require specialized hardware, such as GPUs, FPGAs and network adapters. The enha ...
```
Implemented, thank you!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+However, in Kubernetes v1.32, it is withdrawn and related code will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 
+
+The decision to remove this feature originated from its incompatibility with cluster autoscaling, as resource availability was made non-transparent, complicating decision making for both Cluster Autoscaler and controllers. 
```
Implemented, thanks!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+This enhancement [#3063](https://github.com/kubernetes/enhancements/issues/3063) initially introduced Dynamic Resource Allocation (DRA) in Kubernetes 1.26.
+
+However, in Kubernetes v1.32, it is withdrawn and related code will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 
```
Implemented a slightly different change, thank you!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+## Note on the withdrawal of the old DRA implementation
+
+This enhancement [#3063](https://github.com/kubernetes/enhancements/issues/3063) initially introduced Dynamic Resource Allocation (DRA) in Kubernetes 1.26.
```
Implemented, thanks!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+* Alpha or experimental API versions may be removed in any release without prior deprecation notice, this process can become a withdrawal in cases where a different implementation for the same feature is already in place.
+
+Whether an API is removed as a result of a feature graduating from beta to stable or because that API simply did not succeed, all removals comply with this deprecation policy. Whenever an API is removed, migration options are communicated in the [documentation](/docs/reference/using-api/deprecation-guide/).
```
Implemented, thanks!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+* Beta or pre-release API versions must be supported for 3 releases after the deprecation.
+
+* Alpha or experimental API versions may be removed in any release without prior deprecation notice, this process can become a withdrawal in cases where a different implementation for the same feature is already in place.
```
Done, thanks!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+This blog outlines some of the planned changes for the Kubernetes v1.32 release, that the release team feels you should be aware of, for the continued mantainance of your Kubernetes enviroment and keeping up to date with the latest changes. Information listed below is based on the current status of the v1.32 release and may change before the actual release date. 
+
+### The Kubernetes API removal and deprecation process
+The Kubernetes project has a well-documented [deprecation policy](/docs/reference/using-api/deprecation-policy/) for features. This policy states that stable APIs may only be deprecated when a newer, stable version of that same API is available and that APIs have a minimum lifetime for each stability level. A deprecated API has been marked for removal in a future Kubernetes release, it will conti ...
```
Implemented a slightly different change, thank you!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+As we get closer to the release date for Kubernetes v1.32, the project develops and matures, features may be deprecated, removed, or replaced with better ones for the project's overall health. 
+
+This blog outlines some of the planned changes for the Kubernetes v1.32 release, that the release team feels you should be aware of, for the continued mantainance of your Kubernetes enviroment and keeping up to date with the latest changes. Information listed below is based on the current status of the v1.32 release and may change before the actual release date. 
```
Outdated, already implemented the fixes, but still thank you for pointing these out!

mbianchidev commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+  Rashan Smith
+---
+
+As we get closer to the release date for Kubernetes v1.32, the project develops and matures, features may be deprecated, removed, or replaced with better ones for the project's overall health. 
```
Implemented a slightly different change, thank you!

mbianchidev commented:
> A copy-edit and spelling pass. Feel free to take or leave any given suggestion. Also, I had to refresh mid-review; I've checked for consistency, but I apologize if anything seems misaligned or outdated.

Thank you so much @nate-double-u !
I implemented most of your suggestions, just some slight changes - sorry for the mid-review refresh but I was probably addressing other comments.
It's all good, I really appreciate your contribution and feedback.

stmcginnis commented on content/en/blog/_posts/2024-11-04-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,133 @@
+
+This enhancement [#3063](https://github.com/kubernetes/enhancements/issues/3063) introduced Dynamic Resource Allocation (DRA) in Kubernetes 1.26.
+
+However, in Kubernetes v1.32, such enhancement is withdrawn. Related code will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 
```
```suggestion
However, in Kubernetes v1.32, this approach to DRA has been significantly changed. Code related to the original implementation will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 
```

Just a suggestion on wording, but I do think this paragraph needs to be clarified.

When reading through this, I read:

> such enhancement is withdrawn.

"Oh, DRA has been removed? I didn't realize that." Then:

> Even more DRA enhancements

"Wait, I thought it just said DRA was removed?"

mbianchidev replied:
Good point, implemented your suggestion!

Thank you 😄

mbianchidev commented:
> Hi. Thanks for the PR. A couple of things I'd like to see:
> 
> * use real headings
>   Don't use paragraphs of bold text; use actual headings
> * mention each of the [v1.32 API removals](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-32).
> 
> and some smaller things that are still worth doing
> 
> * write API kinds in UpperCamelCase, per the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/) and _without_ backticks
> * write field names inside backticks
> * Try not to put hyperlinks into headings; if you can reword to avoid this, that's better.
> 
> Because this is an official Kubernetes article, I recommend following our style guide where feasible.

Hey Tim, thanks a lot for your review!

* use real headings
I understand what you say but... The intention I had here was not for that paragraph to have sub-headings. 
The DRA changes are all tightly related to each other and they are meant to be showcased in a single section as they make more sense if read together. I just decided to space the paragraphs and bold the topic of the KEP to improve readability.

Happy to just make it a single paragraph if that's better (still no sub-headings).

* mention each of the API removals
Will do! 

_Even if this is technically not a deprecations and removal blog and I think it would be best to mention it just in the final release blog._

* write API kinds in UpperCamelCase 
* write field names inside backticks
Can help us by pointing out which ones are missing? 
We tried to be careful about it but both me and the previous reviewers might have missed some!

* Try not to put hyperlinks into headings; if you can reword to avoid this, that's better.
What is the rationale behind this? 
Asking because I actually find value in having the KEP link immediately clickable to dive deeper in the enhancement I'm interested in. 
On top of this, a KEP link in a heading has already been done in the previous mid-cycle blog for 1.31, passing review.

We, as Comms, are doing our best to keep this (and every) release article within the style guide and I'm super glad you and the other reviewers are all here to help us.

sftim commented:
> Even if this is technically not a deprecations and removal blog and I think it would be best to just mention it in the final release blog.

If it's _not_ the deprecations and removals article, please open a new PR to add the deprecations and removals article! We should have one.

sftim commented:
About hyperlinks in headings: having them makes the site slightly harder to maintain. We may one day want to make a change that means we first need to clean up all the headings that contain hyperlinks.

mbianchidev commented:
> > Even if this is technically not a deprecations and removal blog and I think it would be best to just mention it in the final release blog.
> 
> If it's _not_ the deprecations and removals article, please open a new PR to add the deprecations and removals article! We should have one.

AFAIK the mid-cycle blog can be either a deprecation and removals, when it makes sense in terms of amount features removed/deprecated or a sneak peek - which is more of a lightweight way to showcase the feature that the release team thinks could be the most interesting for our readers.

CC: @fsmunoz / @katcosgrove / @gracenng - please correct me if I'm wrong here!

IMHO since we only have a withdrawal and one API removal we don't need a full fledged deprecations and removals article.

mbianchidev commented:
> About hyperlinks in headings: having them makes the site slightly harder to maintain. We may one day want to make a change that means we first need to clean up all the headings that contain hyperlinks.

How likely is this change going to happen?
Asking because I really like the idea (as I see value in it) of giving readers easy access to KEP links without venturing in the entire paragraph.

nate-double-u commented:
> > About hyperlinks in headings: having them makes the site slightly harder to maintain. We may one day want to make a change that means we first need to clean up all the headings that contain hyperlinks.
> 
> How likely is this change going to happen? Asking because I really like the idea (as I see value in it) of giving readers easy access to KEP links without venturing in the entire paragraph.

Hyperlinks in headings are rare in the Kubernetes blog (and even rarer in the documentation). Even though links in headers are not explicitly discouraged in the style guide, this can be considered an unwritten rule (Let's consider it written since Tim brought it up?). I don't recommend the practice either if for no other reason than headings become links, and having links in links gets hard to manage. If you need to link something and find the only place to do it is in a header, it's best to rewrite that paragraph to better support what you want to communicate.

kannon92 commented:
The main concern I have with this blog is that I think a lot of the DRA KEPs won't actually make it into 1.32. 

I think dropping classic DRA and promoting structured parameters to beta seem to be looking green for the release. I don't have a lot of confidence that the others will make it into 1.32 at the moment.

mbianchidev commented:
> > > About hyperlinks in headings: having them makes the site slightly harder to maintain. We may one day want to make a change that means we first need to clean up all the headings that contain hyperlinks.
> > 
> > 
> > How likely is this change going to happen? Asking because I really like the idea (as I see value in it) of giving readers easy access to KEP links without venturing in the entire paragraph.
> 
> Hyperlinks in headings are rare in the Kubernetes blog (and even rarer in the documentation). Even though links in headers are not explicitly discouraged in the style guide, this can be considered an unwritten rule (Let's consider it written since Tim brought it up?). I don't recommend the practice either if for no other reason than headings become links, and having links in links gets hard to manage. If you need to link something and find the only place to do it is in a header, it's best to rewrite that paragraph to better support what you want to communicate.
> 
> Edit: I don't know how screen readers would manage links in headers either, so accessibility tools are another argument against this.

The last point is actually *very* important and I didn't think about it!
I just removed all the links from headings.

Let's write the rule in the style guide, tribal knowledge is very dangerous, especially when it relies (supposedly) on a single point of failure.

I don't think trailing links would make a big difference, as headings are anchor links, not fully links (but I could be wrong, not a Hugo expert).
Anyway, it is solved now!

mbianchidev commented:
> The main concern I have with this blog is that I think a lot of the DRA KEPs won't actually make it into 1.32.
> 
> I think dropping classic DRA and promoting structured parameters to beta seem to be looking green for the release. I don't have a lot of confidence that the others will make it into 1.32 at the moment.

Great point, thanks for bringing it up.

The thing is: this blog is due to be published on Monday 4th November and it could be delayed a bit but we have a hard deadline for publishing before KubeCon NA, hard limit at the 9th November.

I would say we cannot afford waiting for exceptions to be filed and eventually KEPs to be removed from the milestone.
Do we just keep the ones we are 100% positive they will be in, risking to be incomplete? (Rethoric question, I'm just thinking out loud here)
Or do we trust that readers will understand that:

> Information listed below is based on the current status of the v1.32 release and may change before the actual release date.

This quote is from the blog itself. Maybe it should be written in a note/warning form.

CC: @fsmunoz - what should we do?

fsmunoz commented:
@mbianchidev thanks for the heads-up.

For context, the "Deprecations and Removals" article has been used as a "sneak peek" at the discretion of the release team - for this cycle we agreed that it would make sense to keep a micycle article, but that considering the absence of major removals, it would follow that format (cf. [1.30 Upcoming Changes](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/)).

The challenge is that, unlike Deprecations and Removals, some of the highlights that we choose are on enhancements that could not make it past Code Freeze. While on the v1.30 example above the midcycle blog was scheduled 1 day after Code Freeze, [for 1.31 it was similar to what we have](https://groups.google.com/a/kubernetes.io/g/dev/c/m_p2IsuDm-Q/m/lqTPs6i2AAAJ), so several days before the Code Freeze.

I see the possible options:

1. Keep the publishing deadline, and go with the ones that we feel are likely to make it.
2. Push the publishing deadline closer to the Code Freeze, but still next week, so that we can have a better view on what's going to make it and adjust if necessary.
3. Push the deadline for after KubeCon NA

I'm very tempted to pull the trigger on 2, which to me is the one that allows us to keep the article in Week 9, while reducing the uncertainty about highlighting things that could be dropped from v1.32. 

- I don't think we should push the deadline for after KubeCon since that would defeat the purpose of the article (from a mid-cycle sneak peek to something that is only a couple of weeks away from the release date). Keeping it to Monday
- I feel that the uncertainty we currently have could be avoided by moving the date closer to Code Freeze, eventually on the same day.

Given that we are close to the publishing date there's not a lot of time to discuss it, but I'll gather feedback and operate on lazy consensus by tomorrow, with an initial bias toward pushing the deadline for the blog to Friday 21:00 UTC, and having a joint review of what could need to be dropped around Wednesday.

On a related note, if DRA enhancements are not ready, I would again ask for exceptions to be opened ASAP so that we can deal with them better, granting more time, etc: the later they are opened, the less ability we have to provide the leeway that could be required (thanks @kannon92 for the broad update on what could make it right now).

cc @katcosgrove @gracenng for awareness.

sftim commented:
> The thing is: this blog is due to be published on Monday 4th November and it could be delayed a bit but we have a hard deadline for publishing before KubeCon NA, hard limit at the 9th November.

Just a note: with the ≅ 4 week mean review time / team commitment, this should've been ideally opened early in October.

mbianchidev commented:
> > The thing is: this blog is due to be published on Monday 4th November and it could be delayed a bit but we have a hard deadline for publishing before KubeCon NA, hard limit at the 9th November.
> 
> Just a note: with the ≅ 4 week mean review time / team commitment, this should've been ideally opened early in October.

Hey Tim, thanks for pointing that out!

As Comms Lead (and previous shadow) I am aware of how the review cycle works, although I think the actual mean review time cold closer to 10-15 days, at least for release blogs as they have been prioritized in the past by sig-docs-blog.
E.g. see the [1.31 deprecation and removals blog](https://github.com/kubernetes/website/pull/46948/commits)

We've been discussing exactly this blog's deadline and the criticality of giving so little to review. We did that on Slack (although not on #release-comms and that's my mistake). 
Unfortunately there was no way that this could've been opened early October as Enhancement Freeze was on the 11th October.
This is the maximum we could do with how the deadlines were defined for v1.32 and going forward I don't see a way to improve on it, it's physiological - for a matter of _consecutio temporum_ - that the mid-cycle blog has to be finalized *after* code freeze.

Being a bit stoic on the topic: any suggestion you might have to move this forward would be greatly appreciated.
Thank you so much!

mbianchidev commented:
> @mbianchidev thanks for the heads-up.
> 
> For context, the "Deprecations and Removals" article has been used as a "sneak peek" at the discretion of the release team - for this cycle we agreed that it would make sense to keep a micycle article, but that considering the absence of major removals, it would follow that format (cf. [1.30 Upcoming Changes](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/)).
> 
> The challenge is that, unlike Deprecations and Removals, some of the highlights that we choose are on enhancements that could not make it past Code Freeze. While on the v1.30 example above the midcycle blog was scheduled 1 day after Code Freeze, [for 1.31 it was similar to what we have](https://groups.google.com/a/kubernetes.io/g/dev/c/m_p2IsuDm-Q/m/lqTPs6i2AAAJ), so several days before the Code Freeze.
> 
> I see the possible options:
> 
> 1. Keep the publishing deadline, and go with the ones that we feel are likely to make it.
> 2. Push the publishing deadline closer to the Code Freeze, but still next week, so that we can have a better view on what's going to make it and adjust if necessary.
> 3. Push the deadline for after KubeCon NA
> 
> I'm very tempted to pull the trigger on 2, which to me is the one that allows us to keep the article in Week 9, while reducing the uncertainty about highlighting things that could be dropped from v1.32.
> 
> * I don't think we should push the deadline for after KubeCon since that would defeat the purpose of the article (from a mid-cycle sneak peek to something that is only a couple of weeks away from the release date). Keeping it to Monday
> * I feel that the uncertainty we currently have could be avoided by moving the date closer to Code Freeze, eventually on the same day.
> 
> Given that we are close to the publishing date there's not a lot of time to discuss it, but I'll gather feedback and operate on lazy consensus by tomorrow, with an initial bias toward pushing the deadline for the blog to Friday 21:00 UTC, and having a joint review of what could need to be dropped around Wednesday.
> 
> On a related note, if DRA enhancements are not ready, I would again ask for exceptions to be opened ASAP so that we can deal with them better, granting more time, etc: the later they are opened, the less ability we have to provide the leeway that could be required (thanks @kannon92 for the broad update on what could make it right now).
> 
> cc @katcosgrove @gracenng for awareness.

Just to update everyone on the topic: we are evalutating the various options on Slack, we will keep everyone posted on which is the option we are aiming for.

mbianchidev commented:
/hold

Until we have a clearer idea about the DRA enhancement making the release, we are holding the mid-cycle blog, ideally until Wednesday, in the worst case scenario we will aim to unhold by Friday.

sftim commented:
Publication date of 2024-11-08 is OK. However, it should only be published if it has been reviewed and approved ahead of publication.

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+For more information, please refer to the [API deprecation guide](/docs/reference/using-api/deprecation-guide/#v1-32).
+
+## Sneak peek of Kubernetes v1.32
+
```
Consider adding introductory text to explain that this is what is _likely_ to be in the release, rather than a commitment.

mbianchidev replied:
Since we are going to delay the publication date to be able to take out the enhancement that won't make the code freeze cut, so the following phrase:

> Information listed below is based on the current status of the v1.32 release and may change before the actual release date.

Should already cover any (minor) change that might occurr after that, wdyt?

sftim replied:
I'd repeat it. Assume that at least some people will skim the article and then publish to social media.

mbianchidev replied:
Done, thanks!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+This enhancement [#3063](https://github.com/kubernetes/enhancements/issues/3063) introduced Dynamic Resource Allocation (DRA) in Kubernetes 1.26.
+
+However, in Kubernetes v1.32, this approach to DRA has been significantly changed. Code related to the original implementation will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 
```
Ideally, rewrite this to make it obvious that Kubernetes v1.32 has yet to be released.

mbianchidev replied:
Done, PTAL.

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816) adds support for a prioritized list of selection criteria attached to a device request in a `ResourceClaim`.
+The DRA's Structured Parameters feature has added the ability to make requests for very specific types of devices using a `ResourceClaim`.
+
+However, the current API did not allow the user to indicate any priority when multiple types or configurations of devices might have met the workload's needs. This feature allows the user to specify alternative requests that satisfy the workloads' need, giving more flexibility to the scheduler in the workloads' scheduling phase.
```
```suggestion
However, the existing API did not allow the user to indicate any priority when multiple types or
configurations of devices might have met the workload's needs. This feature will allow a user
to specify alternative requests that satisfy the workloads' need, giving more flexibility to the
scheduler in the workloads' scheduling phase.
```

mbianchidev replied:
Done!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+**Resource Claim status and standardized network interface data**
+
+KEP [#4817](https://github.com/kubernetes/enhancements/issues/4817) adds driver-owned fields in `ResourceClaim.Status` with the possibility of having standardized network interface data.
+This proposal enhances the `ResourceClaim.Status` field by adding a new sub-field: `Devices`.
+The new entry allows the reporting of driver-specific device status data for each allocated device in a `ResourceClaim`. Allowing the drivers to report such data will improve observability and troubleshooting and enable new functionalities such as network services (in case the IP addresses of a network device are reported).
```
```suggestion
This is a proposal to enhance the `status` field of ResourceClaims by adding a new sub-field: `devices`.
The new field will allow the reporting of driver-specific device status data for each allocated device in
a ResourceClaim. Allowing the drivers to report such data will improve observability and troubleshooting
and enable new functionalities such as network services (in case the IP addresses of a network device are reported).
```

mbianchidev replied:
Implemented, thanks!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815) introduces improvements to Kubernetes' Dynamic Resource Allocation (DRA) framework, enabling more efficient support for high-performance partitionable devices like GPUs, FPGAs, and network adapters. Currently, these devices are treated as whole units, potentially leading to resource inefficiencies when workloads require only a fr ...
+By extending the DRA framework, this enhancement allows workloads to request and consume portions of these resources, increasing utilization and scheduling flexibility. The proposed enhancement also provides primitives for representing both full devices and their partitions more compactly, supporting the recent transition from "classic DRA" to the "structured parameters" approach. 
+The newly proposed extensions, fully transparent to the end-user, enable vendors to provision “overlapping” partitions, ensuring that the scheduler does not allocate conflicting partitions. This approach allows vendors to dynamically create new partitions after allocation, eliminating the need for pre-partitioned configurations while still supporting device partition selection through existing me ...
```
We should avoid assuming or implying that DRA implementations always come from vendors. As a principle of the CNCF, we are open to code being provided gratis by the wider community.

```suggestion
The newly proposed extensions, fully transparent to the end-user, enable device management plugins
to provision _overlapping partitions_, ensuring that the scheduler does not allocate conflicting partitions.
This approach allows device management code to dynamically create new partitions after allocation,
eliminating the need for pre-partitioned configurations while still supporting device partition selection
through existing mechanisms such as ResourceClaim and ResourceSlice (already available in Kubernetes
v1.31).
```

mbianchidev replied:
Implemented, thank you!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+In this release, like the previous one, the Kubernetes project continues introducing enhancements to the Dynamic Resource Allocation (DRA), a key component of the Kubernetes resource management system. These enhancements aim to improve the flexibility and efficiency of resource allocation for workloads that require specialized hardware, such as GPUs, FPGAs and network adapters. The improvements f ...
+
+- Partitionable devices support: [KEP #4815](https://github.com/kubernetes/enhancements/issues/4815)
+- Resource Claim Status and standardized network interface data: [KEP #4817](https://github.com/kubernetes/enhancements/issues/4817)
```
```suggestion
- ResourceClaim status and standardized network interface data: [KEP #4817](https://github.com/kubernetes/enhancements/issues/4817)
```

mbianchidev replied:
Done, thank you!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+- Partitionable devices support: [KEP #4815](https://github.com/kubernetes/enhancements/issues/4815)
+- Resource Claim Status and standardized network interface data: [KEP #4817](https://github.com/kubernetes/enhancements/issues/4817)
+- Structured parameters: [KEP #4381](https://github.com/kubernetes/enhancements/issues/4381)
+- Prioritized Alternatives in Device Requests: [KEP #4816](https://github.com/kubernetes/enhancements/issues/4816)
+- Resource Health Status in Pod Status: [KEP #4680](https://github.com/kubernetes/enhancements/issues/4680)
```
(nit)
```suggestion
- Prioritized alternatives in device requests: [KEP #4816](https://github.com/kubernetes/enhancements/issues/4816)
- Resource sealth status in Pod status: [KEP #4680](https://github.com/kubernetes/enhancements/issues/4680)
```

mbianchidev replied:
Done, thanks!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+### Field status.hostIPs added for Pod
+
+[This enhancement](https://github.com/kubernetes/enhancements/issues/2681) introduces a new field, `status.hostIPs`, to the Kubernetes Pod API, enabling support for multiple IP addresses (IPv4 and IPv6) assigned to a node. Previously, the Pod status only included a singular HostIP field, which limited address configurations, particularly in dual-stack networks. This enhancement allows the API to  ...
+The feature is controlled by the PodHostIPs feature gate, supporting flexibility for gradual rollout.
+
+### Make Kubernetes aware of the LoadBalancer behavior
+
+KEP [#1860](https://github.com/kubernetes/enhancements/issues/1860) graduates to GA, introducing the `ipMode` field for a Service of `type: LoadBalancer`, which can be set to either `"VIP"` or `"Proxy"`. This enhancement is aimed at improving how cloud providers load balancers interact with kube-proxy and it is a change transparent to the end user. The existing behavior of kube-proxy is preserved ...
+
+### Support PSI based on cgroupv2
+[This KEP](https://github.com/kubernetes/enhancements/issues/4205) adds support in the Kubelet to read Pressure Stall Information (PSI) metrics for CPU, Memory and I/O resources exposed from cAdvisor and runc. Pressure metrics are like a barometer that can warn about impending resource shortages on a node. This will enable the Kubelet to report node conditions, which can be utilized to prevent th ...
+
+### Retry generate name for resources
+[This enhancement](https://github.com/kubernetes/enhancements/issues/4420) improves how name conflicts are handled for Kubernetes resources created with the `generateName` field. Previously, if a name conflict occurred, the API server returned a 409 HTTP Conflict error and clients had to manually retry the request. With this update, the API server automatically retries generating a new name up to ...
+
+### Mutating admission policies
+
+[This enhancement](https://github.com/kubernetes/enhancements/issues/3962) introduces "Mutating Admission Policies" using CEL (Common Expression Language), allowing Kubernetes users to define in-line mutating policies without complex admission webhooks. 
+This enhancement simplifies common mutating tasks—such as adding labels, setting fields, or injecting sidecars—by enabling policies to be directly defined within the Kubernetes API server. With CEL's object instantiation and Server-Side Apply merge algorithms, these policies can minimize reinvocation, reduce latency, and eliminate the overhead associated with webhook management. This new approach ...
+
+### Control over terminating Pods in deployments
+
+[This enhancement](https://github.com/kubernetes/enhancements/issues/3973) adds a new field `.spec.podReplacementPolicy` to Deployments to specify whether new Pods can be created only after old Pods have finished terminating or once old Pods start terminating. The default behavior depends on the deployment strategy. Specifying this Pod replacement policy can better control the speed of rollouts/s ...
```
nit: Avoid making "The" or "This" part of the hyperlink.

mbianchidev replied:
Solved, thank you!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+**Add resource health status to the Pod status**
+
+It isn't easy to know when a Pod uses a device that has failed or is temporarily unhealthy. KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680) makes troubleshooting of Pod crashes easier by exposing device health via `Pod Status`.
```
Try this:
```suggestion
It isn't easy to know when a Pod uses a device that has failed or is temporarily unhealthy.
KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680) proposes
exposing device health via Pod `status`, making troubleshooting of Pod crashes easier.
```

mbianchidev replied:
Amazing, implemented!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+**Prioritized alternatives in device requests** 
+
+KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816) adds support for a prioritized list of selection criteria attached to a device request in a `ResourceClaim`.
+The DRA's Structured Parameters feature has added the ability to make requests for very specific types of devices using a `ResourceClaim`.
```
How about:
```suggestion
KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816) proposes support for prioritized
selection criteria associated to a device request in a ResourceClaim.
DRA _structured parameters_ will gain the ability to make requests for very specific types of devices using a ResourceClaim.
```

mbianchidev replied:
Perfect, now I also see what you meant with API Objects withput the backtick. 

A teaching for next blog :)

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+**Partitionable devices support**
+
+KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815) introduces improvements to Kubernetes' Dynamic Resource Allocation (DRA) framework, enabling more efficient support for high-performance partitionable devices like GPUs, FPGAs, and network adapters. Currently, these devices are treated as whole units, potentially leading to resource inefficiencies when workloads require only a fr ...
```
```suggestion
KEP [#4815](https://github.com/kubernetes/enhancements/issues/4815) proposes improvements to
Kubernetes' Dynamic Resource Allocation (DRA) framework, enabling more efficient support for
high-performance partitionable devices such as GPUs, FPGAs, and network adapters.
Currently, these devices are treated as whole units, potentially leading to resource inefficiencies when
workloads require only a fraction of a device’s capacity. 
```
?

mbianchidev replied:
Yep, good idea!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+With the graduation of [this feature](https://github.com/kubernetes/enhancements/issues/4369) to beta, Kubernetes now allows almost all printable ASCII characters (excluding "=") to be used as environment variable names. This change addresses the limitations previously imposed on variable naming, facilitating a broader adoption of Kubernetes by accommodating various application needs. The relaxed ...
+
+### Field status.hostIPs added for Pod
+
+[This enhancement](https://github.com/kubernetes/enhancements/issues/2681) introduces a new field, `status.hostIPs`, to the Kubernetes Pod API, enabling support for multiple IP addresses (IPv4 and IPv6) assigned to a node. Previously, the Pod status only included a singular HostIP field, which limited address configurations, particularly in dual-stack networks. This enhancement allows the API to  ...
```
Optionally, link to the concept page about the downward API

mbianchidev replied:
Cool, linked it!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+The original dynamic resource allocation (DRA) uses claim and class parameters that are opaque to Kubernetes. KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) introduces structured parameters so that kube-scheduler and Cluster Autoscaler can handle a claim allocation without relying on a third-party driver.
+As users increasingly adopt Kubernetes as their management solution for batch-processing workloads and edge computing, the need to expose specialized hardware to Pods is increasing.
+Such workloads no longer need just RAM and CPU, so this KEP introduced a new API to describe which of these specialized resources a Pod needs, such as devices like a GPU or other kinds of accelerators.
```
```suggestion
Such workloads no longer need just RAM and CPU, so Kubernetes v1.32 will introduce a new API to describe
which of these specialized resources a Pod needs, such as devices like a GPU or other kinds of accelerators.
The DRA framework allows you to provide the extra resources dynamically, rather than configuring them
as static resources that are available whether or not any Pod requires them.
```
?

mbianchidev replied:
Better, implemented it!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+By extending the DRA framework, this enhancement allows workloads to request and consume portions of these resources, increasing utilization and scheduling flexibility. The proposed enhancement also provides primitives for representing both full devices and their partitions more compactly, supporting the recent transition from "classic DRA" to the "structured parameters" approach. 
+The newly proposed extensions, fully transparent to the end-user, enable vendors to provision “overlapping” partitions, ensuring that the scheduler does not allocate conflicting partitions. This approach allows vendors to dynamically create new partitions after allocation, eliminating the need for pre-partitioned configurations while still supporting device partition selection through existing me ...
+
+**Resource Claim status and standardized network interface data**
```
```suggestion
**ResourceClaim status and standardized network interface data**
```

mbianchidev replied:
Solved!

sftim commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+**Resource Claim status and standardized network interface data**
+
+KEP [#4817](https://github.com/kubernetes/enhancements/issues/4817) adds driver-owned fields in `ResourceClaim.Status` with the possibility of having standardized network interface data.
```
```suggestion
KEP [#4817](https://github.com/kubernetes/enhancements/issues/4817) proposes adding new
driver-owned fields into the `status` of a ResourceClaim, including the new possibility of publishing
standardized network interface data.
```

mbianchidev replied:
Done!

mbianchidev commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+The new entry allows the reporting of driver-specific device status data for each allocated device in a `ResourceClaim`. Allowing the drivers to report such data will improve observability and troubleshooting and enable new functionalities such as network services (in case the IP addresses of a network device are reported).
+
+
+**DRA Structured Parameters** 
```
Didn't see this one, fixed!

mbianchidev commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+### Mutating admission policies
+
+[This enhancement](https://github.com/kubernetes/enhancements/issues/3962) introduces "Mutating Admission Policies" using CEL (Common Expression Language), allowing Kubernetes users to define in-line mutating policies without complex admission webhooks. 
```
Way better, implemented it!

mbianchidev commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+
+### Allow special characters in environment variables
+
+With the graduation of [this feature](https://github.com/kubernetes/enhancements/issues/4369) to beta, Kubernetes now allows almost all printable ASCII characters (excluding "=") to be used as environment variable names. This change addresses the limitations previously imposed on variable naming, facilitating a broader adoption of Kubernetes by accommodating various application needs. The relaxed ...
```
Implemented, thanks!

mbianchidev commented:
> Publication date of 2024-11-08 is OK. However, it should only be published if it has been reviewed and approved ahead of publication.

Thank you for your review Tim, I addressed most comments, PTAL.

Of course we aim for approval and LGTM before publication, that is a given and we are not trying to avoid the proper review process here but we need to get prioritized as unfortunately the deadlines we had set did not allow this blog to be drafted 30 days in advance as it is desirable for reviewers.
I will add a note in the Release Comms handbook specifying that the date of publication of the mid-cycle blog has to be at least 4/5 days *AFTER* code freeze, to avoid incurring in this situation again.

FYI, I also sent a message to inform the current PR Wrangler and the SIG Docs Chairs about the change of deadline and the strategy we are going for, you can find it [here](https://kubernetes.slack.com/archives/CJDHVD54J/p1730631590708249).

We are doing our best to ensure this blog content is actual and useful for our readers and we thank every reviewer that helped us so far!

fsmunoz commented:
As an additional, https://github.com/kubernetes/enhancements/issues/4945 could be well-worth a mention, considering that 100% conformance testing coverage is an important milestone for the release.

mbianchidev commented:
> As an additional, [kubernetes/enhancements#4945](https://github.com/kubernetes/enhancements/issues/4945) could be well-worth a mention, considering that 100% conformance testing coverage is an important milestone for the release.

As discussed privately, to avoid adding further uncertainty to this specific piece, we are going to announce this (if it makes the cut) on the final release blog.

dipesh-rawat commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+The decision to remove this feature originated from its incompatibility with cluster autoscaling, as resource availability was made non-transparent, complicating decision-making for both Cluster Autoscaler and controllers. 
+The newly added Structured Parameter model substitutes the functionality.
+
+This removal will Kubernetes to handle new hardware requirements and resource claims more predictably, bypassing the complexities of back and forth API calls to the kube-apiserver.
```
```suggestion
This removal will allow Kubernetes to handle new hardware requirements and resource claims more predictably, bypassing the complexities of back and forth API calls to the kube-apiserver.
```

dipesh-rawat commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,143 @@
+This [enhancement](https://github.com/kubernetes/enhancements/issues/2681) introduces a new field, `status.hostIPs`, to the Kubernetes Pod API, enabling support for multiple IP addresses (IPv4 and IPv6) assigned to a node. Previously, the Pod status only included a singular HostIP field, which limited address configurations, particularly in dual-stack networks. This enhancement allows the API to  ...
+The feature is controlled by the PodHostIPs feature gate, supporting flexibility for gradual rollout.
+
+For more information about giving containers information about themselves, visit the [downward-api documentation](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/).
```
Use a site-relative link here rather than an absolute URL.

```suggestion
For more information about giving containers information about themselves, visit the [downward-api documentation](/docs/concepts/workloads/pods/downward-api/).
```

mbianchidev commented:
Removed the enhancements not tracked for code freeze as agreed @fsmunoz, seeking approve and LGTM labels from @kubernetes/sig-docs-blog-owners

divya-mohan0209 commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,92 @@
+
+## Note on the withdrawal of the old DRA implementation
+
+This enhancement [#3063](https://github.com/kubernetes/enhancements/issues/3063) introduced Dynamic Resource Allocation (DRA) in Kubernetes 1.26.
```
Nit.
```suggestion
The enhancement [#3063](https://github.com/kubernetes/enhancements/issues/3063) introduced Dynamic Resource Allocation (DRA) in Kubernetes 1.26.
```

OR 
```suggestion
KEP [#3063](https://github.com/kubernetes/enhancements/issues/3063) introduced Dynamic Resource Allocation (DRA) in Kubernetes 1.26.
```

divya-mohan0209 commented on content/en/blog/_posts/2024-11-08-kubernetes-1.32-sneak-peek.md:
```diff
@@ -0,0 +1,92 @@
+
+However, in Kubernetes v1.32, this approach to DRA will be significantly changed. Code related to the original implementation will be removed, leaving KEP [#4381](https://github.com/kubernetes/enhancements/issues/4381) as the "new" base functionality. 
+
+The decision to remove this feature originated from its incompatibility with cluster autoscaling, as resource availability was made non-transparent, complicating decision-making for both Cluster Autoscaler and controllers. 
```
Nit: The sentence might read to some as though DRA is being removed and would benefit from rewording.  

I've suggested something below. WDYT?

```suggestion
The decision to change the existing approach originated from its incompatibility with cluster autoscaling as resource availability was non-transparent, complicating decision-making for both Cluster Autoscaler and controllers. 
```

divya-mohan0209 reviewed (changes requested):
Overall LGTM! Thank you for your work on this, @mbianchidev :) 

Left some suggestions to improve overall clarity, and we should be good to go, IMO.

mbianchidev commented:
Thank you so much @divya-mohan0209 for this last minute review! 

Addressed your comments, PTAL.

divya-mohan0209 reviewed (commented):
Pinging @natalisucks for the final approve!
/lgtm

natalisucks commented:
thanks everyone for your great work on this!
/hold cancel
/approve

fsmunoz commented:
Looking great, thank you all as well!
